### PR TITLE
Add Week6

### PR DESCRIPTION
**Pull request by MarcRaeder**
Add Week6 with following files;
- world-cup
- sentimental-hello
- sentimental-mario-less
- sentimental-mario-more
- sentimental-cash
- sentimental-credit
- sentimental-readability
- dna

### CODE REVIEW

File: week6/dna/dna.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    with open(sys.argv[1]) as file:
+        reader = csv.DictReader(file)
+        database = list(reader)
+
+    with open(sys.argv[2]) as file2:
+        sequence = file2.read()
+
+    matches = {}
+    for i in database[0]:
+        matches[i] = (longest_match(sequence, i))
```
Comment (stefanluth) [inline week6/dna/dna.py]:
```suggestion
        matches[i] = longest_match(sequence, i)
```

**Comment by stefanluth** [comment]
Could you add [typing](https://docs.python.org/3/library/typing.html)?

File: week6/dna/dna.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+            suspect = database[i]['name']
+            break
+        else:
+            suspect_counter = 1
+
+    print(suspect)
+
+    return
+
+def longest_match(sequence, subsequence):
```
Comment (stefanluth) [inline week6/dna/dna.py]:
```suggestion
def longest_match(sequence: list[str], subsequence: list[str]) -> str:
```

Not sure if these are the correct types but this is an example of how to use them

File: week6/world-cup/tournament.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+def simulate_round(teams):
+    """Simulate a round. Return a list of winning teams."""
+    winners = []
+
+    # Simulate games for all pairs of teams
+    for i in range(0, len(teams), 2):
+        if simulate_game(teams[i], teams[i + 1]):
+            winners.append(teams[i])
+        else:
+            winners.append(teams[i + 1])
```
Comment (matthiasdanne) [inline week6/world-cup/tournament.py]:
```suggestion
        first_team_won = simulate_game(teams[i], teams[i + 1])
        if first_team_won:
            winners.append(teams[i])
        else:
            winners.append(teams[i + 1])
```

File: week6/world-cup/tournament.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    for team in sorted(counts, key=lambda team: counts[team], reverse=True):
+        print(f"{team}: {counts[team] * 100 / N:.1f}% chance of winning")
+
+
+def simulate_game(team1, team2):
+    """Simulate a game. Return True if team1 wins, False otherwise."""
+    rating1 = team1["rating"]
+    rating2 = team2["rating"]
+    probability = 1 / (1 + 10 ** ((rating2 - rating1) / 600))
+    return random.random() < probability
```
Comment (matthiasdanne) [inline week6/world-cup/tournament.py]:
```suggestion
    probability = 1 / (1 + 10 ** ((rating2 - rating1) / 600))
    
    return random.random() < probability
```

File: week6/sentimental-readability/readability.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+for i in text:
+    if i.isalpha():
+        letters += 1
+    elif i == " ":
+        words += 1
+    elif i == "." or i == "!" or i == "?":
+        sentences += 1
+
+index = 0.0588 * (letters / words * 100) - 0.296 * \
```
Comment (matthiasdanne) [inline week6/sentimental-readability/readability.py]: `index` is a rather generic name, you should name the used index here. 
If a reviewer wants further information on how or why this index is calculated he/she has a term to google. (And can compare wether this implementation of the used index is correct.)

```suggestion
coleman_liau_index = 0.0588 * (letters / words * 100) - 0.296 * \
```

File: week6/sentimental-mario-more/mario.py [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+        Height = int(input("Height: "))
+    except ValueError:
+        print("Height have to be an integer!")
+    if Height < 9 and Height > 0:
+        break
+    else:
+        print("Height have to be between 1 - 8")
+
+for i in range(Height):
+    print(" " * (Height - i), "#" * (i + 1), " ", "#" * (i + 1))
```
Comment (matthiasdanne) [inline week6/sentimental-mario-more/mario.py]: Oneliners are cool, yeah.

It is however debatable wether a longer solution would increase readability. In this rather simple task i would say no.

**Review by matthiasdanne — CHANGES_REQUESTED** [review]
Haven't reviewed credit & dna so far.

Rest looks good, i have requested minor changes. 👍

File: week6/dna/dna.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+    with open(sys.argv[1]) as file:
+        reader = csv.DictReader(file)
+        database = list(reader)
+
+    with open(sys.argv[2]) as file2:
+        sequence: str = file2.read()
+
+    matches = {}
+    for i in database[0]:
```
Comment (stefanluth) [inline week6/dna/dna.py]:
```suggestion
    for subsequence in database[0]:
```

File: week6/sentimental-cash/cash.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
+from shelve import DbfilenameShelf
```
Comment (stefanluth) [inline week6/sentimental-cash/cash.py]: What's this?

File: week6/sentimental-credit/credit.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    if visa == 4 and (lenght == 13 or lenght == 16):
+        print("VISA")
+    elif lenght == 15 and (amex == 34 or amex == 37):
+        print("AMEX")
+    elif lenght == 16 and (51 <= master <= 55):
+        print("MASTERCARD")
+    else:
+        print("INVALID")
+else:
+    print("INVALID")
```
Comment (stefanluth) [inline week6/sentimental-credit/credit.py]: Move to a `main()` function

File: week6/sentimental-credit/credit.py [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    odd_digits = digits[-1:: -2]
+    even_digits = digits[-2:: -2]
+    checksum = 0
+    checksum += sum(odd_digits)
+    for d in even_digits:
+        checksum += sum(digits_of(d * 2))
+    return checksum % 10
+
+
+lenght = 0
```
Comment (stefanluth) [inline week6/sentimental-credit/credit.py]:
```suggestion
length = 0
```

File: week6/sentimental-credit/credit.py [hunk 11]
```diff
@@ -0,0 +1,46 @@
+while True:
+    card = int(input("Card: "))
+    if card > 0:
+        break
```
Comment (stefanluth) [inline week6/sentimental-credit/credit.py]: Add try/catch similar to `cash.py`

Screenshot 2022-08-25 at 16 31 00

File: week6/sentimental-credit/credit.py [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+def luhn_checksum(card:int) -> int:
+    def digits_of(n):
+        return [int(d) for d in str(n)]
+    digits = digits_of(card)
+    odd_digits = digits[-1:: -2]
+    even_digits = digits[-2:: -2]
+    checksum = 0
+    checksum += sum(odd_digits)
+    for d in even_digits:
```
Comment (stefanluth) [inline week6/sentimental-credit/credit.py]:
```suggestion
    checksum += sum(odd_digits)

    for d in even_digits:
```

File: week6/sentimental-credit/credit.py (hunk 9 above)
Comment (stefanluth) [inline week6/sentimental-credit/credit.py]:
```suggestion
else:
    print("INVALID")
```

Redundant

File: week6/sentimental-mario-less/mario.py [hunk 13]
```diff
@@ -0,0 +1,12 @@
+while True:
+    try:
+        Height = int(input("Height: "))
```
Comment (stefanluth) [inline week6/sentimental-mario-less/mario.py]:
```suggestion
        height = int(input("Height: "))
```

File: week6/sentimental-mario-more/mario.py [hunk 14]
```diff
@@ -0,0 +1,12 @@
+while True:
+    try:
+        Height = int(input("Height: "))
```
Comment (stefanluth) [inline week6/sentimental-mario-more/mario.py]:
```suggestion
        height = int(input("Height: "))
```

File: week6/world-cup/tournament.py [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+
+    return winners
+
+
+def simulate_tournament(teams:list) -> list:
+    """Simulate a tournament. Return name of winning team."""
+    while len(teams) > 1:
+        teams = simulate_round(teams)
+
+    return teams[0]["team"]
```
Comment (stefanluth) [inline week6/world-cup/tournament.py]: Is the return type really `list`?

**Review by stefanluth — CHANGES_REQUESTED** [review]
Use 

```python
if __name__ == "__main__":
    main()
```

in all files :)

File: week6/sentimental-credit/credit.py [hunk 16]
```diff
@@ -0,0 +1,46 @@
+while True:
+    card = int(input("Card: "))
+    if card > 0:
+        break
+
+
+def luhn_checksum(card:int) -> int:
```
Comment (stefanluth) [inline week6/sentimental-credit/credit.py]:
```suggestion
def luhn_checksum(card: int) -> int:
```

File: week6/world-cup/tournament.py [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+def simulate_game(team1: int, team2: int) -> int: 
+    """Simulate a game. Return True if team1 wins, False otherwise."""
+    rating1 = team1["rating"]
+    rating2 = team2["rating"]
+    probability = 1 / (1 + 10 ** ((rating2 - rating1) / 600))
+    
+    return random.random() < probability
+
+
+def simulate_round(teams:list[str]) -> list[str]:
```
Comment (stefanluth) [inline week6/world-cup/tournament.py]:
```suggestion
def simulate_round(teams: list[str]) -> list[str]: 
```

File: week6/world-cup/tournament.py [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+        first_team_won = simulate_game(teams[i], teams[i + 1])
+        if first_team_won:
+            winners.append(teams[i])
+        else:
+            winners.append(teams[i + 1])
+
+    return winners
+
+
+def simulate_tournament(teams:list) -> list:
```
Comment (stefanluth) [inline week6/world-cup/tournament.py]:
```suggestion
def simulate_tournament(teams: list) -> list:
```

**Review by stefanluth — APPROVED** [review]
Nice

**Review by matthiasdanne — APPROVED** [review]
Cool. 👍

**Pull request merged**